RFC: Preparation for Julia 0.6

andreasnoack opened the pull request:
The plan is to remove that formula and modeling functionality and hopefully `StatsModels` will be able to handle that part. For now, I've commented out the formula and modeling related tests. To make this easier, I've loosened the requirement on the `Vector` input to the `DataFrame` constructors such that the element type doesn't have to be `Any`. With that change, the `ModelFrame` method in `StatsModels` works.

There are also various routine syntax updates that should be uncontroversial.

Finally, I plan to remove the dependency on Juno. It was controversial when added and has caused problems since added. E.g., right now `Juno` breaks the REPL and this is not captured by `Juno`'s tests. Getting nice printing of `DataFrames` in `Juno` must be handled elsewhere.

nalimilan commented:
Too bad I just made the same changes in DataTables (https://github.com/JuliaData/DataTables.jl/pull/9, https://github.com/JuliaData/DataTables.jl/pull/8; FWIW the latter has a few more fixes). Maintaining the two repos is parallel leads to some waste of time...

One issue is whether StatsModels is going to be able to support both DataFrames and DataTables at the same time. We likely need the three of them to use a common AbstractTable interface for that. Until then, I'd rather keep the modeling functions in DataFrames.

Is the Juno dep a problem on 0.6?

nalimilan commented on src/subdataframe/subdataframe.jl:
```diff
@@ -51,7 +51,7 @@ immutable SubDataFrame{T <: AbstractVector{Int}} <: AbstractDataFrame
     rows::T # maps from subdf row indexes to parent row indexes
 
-    function SubDataFrame(parent::DataFrame, rows::T)
+    @compat function SubDataFrame{T}(parent::DataFrame, rows::T) where T
```
Is `T` really needed here?

nalimilan replied:
Nevermind, I just found the notes in https://github.com/JuliaLang/julia/pull/20308/files.

nalimilan replied:
Yet, shouldn't `T<:AbstractVector{Int}`? Not sure what would happen without it, probably an error a bit later. In his PR Jeff added these (duplicated) constraints.

andreasnoack replied:
Not sure. I just followed the deprecation warning. Shouldn't harm to make that restriction though.

andreasnoack commented:
> Too bad I just made the same changes in DataTables (JuliaData/DataTables.jl#9, JuliaData/DataTables.jl#8; FWIW the latter has a few more fixes). Maintaining the two repos is parallel leads to some waste of time...

I agree. That is why I've been asking all the questions about `DataArrays`/`NullableArrays` lately. Ideally, we didn't maintain two almost identical repos in parallel but I find `DataArrays` so much more convenient to work with that I'm willing to contribute time for now. I've tried to find all the best arguments for `NullableArrays` to convince myself that moving over is a good idea but so far I'm not convinced that it is an improvement relative to `DataArrays`.

> Until then, I'd rather keep the modeling functions in DataFrames.

I concluded that it wasn't really feasible because `~` is not a macro anymore. Fixing the formula code here would therefore require reimplementing the formula code from `StatsModels`. I tried to use `StatsModels` with `DataFrames` and it mostly worked. Many of the methods can dispatch on `DataFrame`/`DataTable` and do the right thing. However, I need your PRs merged into `DataTables` before I can test this further.

> Is the Juno dep a problem on 0.6?

Yes. If you want to use the REPL 😄 (see my last comment). Juno is not sufficiently tested to capture these bugs so it is up to us to detect that `using Juno` breaks the REPL. But even if that was fixed I think it is the wrong dependency direction. If Juno doesn't want to depend on this package then we'd be better off with a `JunoDataFrames`. When we have conditional modules, that package could be integrated into Juno.

nalimilan commented:
> I tried to use StatsModels with DataFrames and it mostly worked. Many of the methods can dispatch on DataFrame/DataTable and do the right thing. However, I need your PRs merged into DataTables before I can test this further.

That's cool, but we still need to define the functions in a separate interface package so that StatsModels can use them without DataFrames and DataTables conflicting. AbstractTables is the most logical candidate, but I'm not sure how stable is its current API.

I'm not really in favor of keeping the Juno code, I just wanted to know whether it was one of the reasons for the test failures on 0.6.

nalimilan commented on test/join.jl:
```diff
@@ -18,9 +18,9 @@ module TestJoin
-    left = outer[!isna(outer[:Name]), :]
-    inner = left[!isna(left[:Job]), :]
+    right = outer[.!isna(outer[:Job]), [:Name, :ID, :Job]]
+    left = outer[.!isna(outer[:Name]), :]
```
Will that work on 0.5? I would try with `(!).(...)`.

andreasnoack replied:
No it won't. `(!).(...)` is indeed necessary. Just realized that while fixing another package.

andreasnoack commented:
Updated. I've dropped 0.4 support, deleted the statsmodels and formula code but kept the contrasts code. However, the contrasts tests rely on `ModelFrame`s for now, so I've disabled the tests. Depending on the StatsModels solution the contrasts could be dropped from this package or we could change the tests to avoid `ModelFrames`.

nalimilan commented on src/subdataframe/subdataframe.jl:
```diff
@@ -51,14 +51,14 @@ immutable SubDataFrame{T <: AbstractVector{Int}} <: AbstractDataFrame
     rows::T # maps from subdf row indexes to parent row indexes
 
-    @compat function SubDataFrame{T}(parent::DataFrame, rows::T) where T
+    function (::Type{SubDataFrame{T}}){T}(parent::DataFrame, rows::T)
```
Is this the new recommended syntax for inner constructors? I'm still confused about this, and the manual doesn't help.

andreasnoack replied:
I'm sure it is recommended but it works without warnings across 0.5 and 0.6 so I think it the right solution until we drop 0.5 support.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -490,19 +490,19 @@ complete_cases!(df::AbstractDataFrame)
 
 """
-complete_cases!(df::AbstractDataFrame) = deleterows!(df, find(!complete_cases(df)))
+completecases!(df::AbstractDataFrame) = deleterows!(df, find(!completecases(df)))
```
Better do `find(!, completecases(df))` as in DataTables to avoid allocating a copy.

nalimilan commented on src/abstractdataframe/io.jl:
```diff
@@ -181,7 +181,7 @@ end
     write(io, "</thead>")
     write(io, "<tbody>")
-    tty_rows, tty_cols = _displaysize(io)
+    tty_rows, tty_cols = Base.displaysize(io)
```
No need for `Base.` here and below.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -292,7 +292,6 @@ function contrasts_matrix(C::HelmertCoding, baseind, n)
     mat = mat[[baseind; 1:(baseind-1); (baseind+1):end], :]
     return mat
 end
-    
```
?

ararslan replied:
Removes trailing whitespace. Note that the diff shows that this line had the whitespace removed in favor of an empty line below.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -580,27 +580,27 @@ end
-end
-
-@render Inline df::AbstractDataFrame _render(df)
+# using Juno
```
Just remove this altogether, that's easy to find in the git history if needed.

nalimilan commented:
I remove all of the code related to modeling, or keep all of it. I don't like commenting out code.

Other than that, looks good to me.

nalimilan commented on test/runtests.jl:
```diff
@@ -16,17 +16,16 @@ my_tests = ["utils.jl",
-            "statsmodel.jl",
-            "contrasts.jl"]
+            "show.jl"#,
+            # "contrasts.jl"
```
Remove this line and I would say it's good to go.

nalimilan replied:
(And remove the corresponding file...)

ararslan commented:
Until StatsModels works flawlessly with DataFrames, I think we should just copy the definition of `@formula` from StatsModels here. None of the other modeling code in this package would need to change, just the definition of formulas.

andreasnoack commented:
The right solution might then be to make an update for 0.5 before these fixes where we deprecate the use of `~` and clearly write how to transition, i.e. use StatsModels. However, it requires that we make StatsModels work generically first. Hopefully soon.

ararslan commented:
That sounds good to me. We should start an issue to track everything that's needed to get StatsModels working generically, or at least across DataFrames and DataTables.

cjprybol commented on src/deprecated.jl:
```diff
@@ -9,12 +9,10 @@ import Base: keys, values, insert!
 
 @deprecate read_rda(args...) FileIO.load(args...)
+
+@deprecate complete_cases(df)  complete_cases(df)
```
```julia
@deprecate complete_cases(df)  completecases(df)
```

cjprybol commented on src/deprecated.jl:
```diff
@@ -9,12 +9,10 @@ import Base: keys, values, insert!
 @deprecate read_rda(args...) FileIO.load(args...)
+
+@deprecate complete_cases(df)  complete_cases(df)
+@deprecate complete_cases!(df) complete_cases!(df)
```
```julia
@deprecate complete_cases!(df) completecases!(df)
```

cjprybol replied:
given https://github.com/JuliaData/DataTables.jl/pull/6#issue-207690988 and linked issues could this be changed to deprecate `complete_cases!(df)` to `dropna!(df)`?

would need `dropna` too I guess https://github.com/JuliaData/DataTables.jl/pull/6#issuecomment-280668091

nalimilan commented:
> That sounds good to me. We should start an issue to track everything that's needed to get StatsModels working generically, or at least across DataFrames and DataTables.

https://github.com/JuliaStats/StatsModels.jl/issues/14

ararslan commented:
Heh, whoops. Missed that. Thanks!

davidanthoff commented:
Bump, what is the status here? Would be great if we had a tagged version of DataFrames that works on julia 0.6. I'm trying to move ExcelReaders (and then Query) to 0.6, but my CI builds won't pass until there is a tagged version of DataFrames that works on julia 0.6.

ararslan commented:
DataArrays master works on 0.6 so once that's tagged, I think that version of DataArrays + this PR = a complete working DataFrames on 0.6.

nalimilan commented:
Though model formula support still isn't there AFAIK. I guess we could still tag a beta version without it to fix packages which don't need formulas (most of them).

andreasnoack commented:
I've been traveling and busy but I hope to have time work a little on this again next week.

davidanthoff commented:
I think it would be great if there was a version tagged with a REQUIRE line ``julia 0.6-`` really soon, even if that version still has some known issues. Right now a simple ``using DataFrames`` breaks on julia 0.6, and that pretty much means that a huge number of packages can't even start to work on 0.6 compatibility without going through a lot of hoops.

ararslan commented:
I've dealt with the formulas situation temporarily in #1170 by copying over the requisite definitions from StatsModels. That way we have something here that will work in the interim while StatsModels is made to work with DataFrames.
